### PR TITLE
add new columns of wwpn in fcp table

### PR DESCRIPTION
**Pull request by SharpRazor**
Signed-off-by: cao biao <bjcb@cn.ibm.com>

### CODE REVIEW

File: zvmsdk/database.py [hunk 1]
```diff
@@ -43,6 +43,9 @@
 _DBLOCK_IMAGE = threading.RLock()
 _DBLOCK_GUEST = threading.RLock()
 _DBLOCK_FCP = threading.RLock()
+# new added columns from CIC V115
+FCP_NEW_ADDED_COLS = [dict(col_name='wwpn_npiv', col_type='char(16)'),
```
Comment (dongyanyang) [inline zvmsdk/database.py:47]: why here use char(16) and the _initialize_table use varchar(16)?

File: zvmsdk/database.py [hunk 2]
```diff
@@ -287,10 +291,23 @@ def _initialize_table(self):
...
+            'wwpn_phy       varchar(16) COLLATE NOCASE)'))
         with get_fcp_conn() as conn:
             conn.execute(sql)
 
+    def _upgrade_table(self):
+        """Add columns that do not exist in older version."""
+        with get_fcp_conn() as conn:
+            for col in FCP_NEW_ADDED_COLS:
+                try:
+                    conn.execute("ALTER TABLE fcp ADD COLUMN %s %s" %
```
Comment (dongyanyang) [inline zvmsdk/database.py:305]: why don't you first check whether the columns already exist?

File: zvmsdk/database.py [hunk 3]
```diff
@@ -287,10 +291,23 @@ def _initialize_table(self):
...
+                                 (col['col_name'], col['col_type']))
+                except sqlite3.OperationalError as e:
+                    LOG.warning('When upgrading table FCP, encounter '
```
Comment (dongyanyang) [inline zvmsdk/database.py:308]: I think we should not ignore upgrade error.

File: zvmsdk/volumeop.py [hunk 4]
```diff
@@ -919,6 +919,11 @@ def _sync_db_with_zvm(self):
                 LOG.warn("FCP {} not found in ZVM.".format(fcp))
                 comment_dict['state'] = 'notfound'
             self.db.update_comment_of_fcp(fcp, comment_dict)
+            # update wwpns
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:922]: Should only update the wwpns when it is not already in DB.
So, you need to update the get_fcp_dict_in_db to also return the WWPNs info from DB.

File: zvmsdk/volumeop.py [hunk 5]
```diff
@@ -1248,8 +1253,20 @@ def get_volume_connector(self, assigner_id, reserve):
...
+                if wwpn_npiv:
+                    # found wwpn in database
+                    LOG.warning("FCP device %s not found in zvm, but got its "
+                                "NPIV wwpn %s in database." % (fcp_no,
+                                                               wwpn_npiv))
+                    wwpns.append(wwpn_npiv)
+                else:
+                    # if wwpn also not exist in db, log errors
+                    errmsg = "FCP device %s has no available WWPN." % fcp_no
+                    LOG.error(errmsg)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1269]: line 1278 should also be updated.

**Review by dongyanyang — CHANGES_REQUESTED** [review]

**Comment by SharpRazor** [comment]
## test results
### 1. drop fcp table in fcp database
### 2. restart sdkserver
### 3. go into the FCP database:
```
sqlite> select * from fcp;
fcp_id|assigner_id|connections|reserved|path|comment|wwpn_npiv|wwpn_phy
5c0a||0|0|0|{'state': 'free'}|c05076ddf7003c00|c05076ddf7001181
5c07||0|0|0|{'state': 'free'}|c05076ddf7003bf4|c05076ddf7001181
5d0b||0|0|1|{'state': 'free'}|c05076ddf700411c|c05076ddf7001d81
5d0f||0|0|1|{'state': 'free'}|c05076ddf700412c|c05076ddf7001d81
5c09||0|0|0|{'state': 'free'}|c05076ddf7003bfc|c05076ddf7001181
5c0f||0|0|0|{'state': 'free'}|c05076ddf7003c14|c05076ddf7001181
5c06||0|0|0|{'state': 'free'}|c05076ddf7003bf0|c05076ddf7001181
5c0e||0|0|0|{'state': 'free'}|c05076ddf7003c10|c05076ddf7001181
5d0a||0|0|1|{'state': 'free'}|c05076ddf7004118|c05076ddf7001d81
5c0d||0|0|0|{'state': 'active'}|c05076ddf7003c0c|c05076ddf7001181
5d09||0|0|1|{'state': 'active'}|c05076ddf7004114|c05076ddf7001d81
5d0d||0|0|1|{'state': 'active'}|c05076ddf7004124|c05076ddf7001d81
5c08||0|0|0|{'state': 'active'}|c05076ddf7003bf8|c05076ddf7001181
5c0c||0|0|0|{'state': 'free'}|c05076ddf7003c08|c05076ddf7001181
5d07||0|0|1|{'state': 'free'}|c05076ddf700410c|c05076ddf7001d81
5d08||0|0|1|{'state': 'free'}|c05076ddf7004110|c05076ddf7001d81
5d0c||0|0|1|{'state': 'free'}|c05076ddf7004120|c05076ddf7001d81
5d06||0|0|1|{'state': 'free'}|c05076ddf7004108|c05076ddf7001d81
5d0e||0|0|1|{'state': 'free'}|c05076ddf7004128|c05076ddf7001d81
5c0b||0|0|0|{'state': 'free'}|c05076ddf7003c04|c05076ddf7001181
sqlite> 
```

File: zvmsdk/database.py [hunk 6]
```diff
@@ -380,11 +383,13 @@ def get_all_fcps_of_assigner(self, assigner_id=None):
...
                 result = conn.execute("SELECT fcp_id, assigner_id, "
-                        "connections, reserved, path, comment FROM "
-                        "fcp WHERE assigner_id=?", (assigner_id,))
+                                      "connections, reserved, path, comment, "
+                                      "wwpn_npiv, wwpn_phy FROM fcp WHERE "
+                                      "assigner_id=?", (assigner_id,))
             else:
                 result = conn.execute("SELECT fcp_id, assigner_id, "
-                        "connections, reserved, path, comment FROM fcp")
+                                      "connections, reserved, path, comment, "
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/database.py:394]: done.

**Review by dongyanyang — APPROVED** [review]

**Review by bjhuangr — APPROVED** [review]

**Pull request merged**